allauth/socialaccount/adapter.py

```diff
@@ -197,9 +197,22 @@ def get_app(self, request, provider):
 
         config = app_settings.PROVIDERS.get(provider, {}).get("APP")
         if config:
-            app = SocialApp(provider=provider)
-            for field in ["client_id", "secret", "key", "certificate_key"]:
-                setattr(app, field, config.get(field))
+            if (
+                app_settings.STORE_TOKENS
+                and SocialApp.objects.filter(provider=provider).exists()
+            ):
+                # We already saved the app. Fetch and return it right away...
+
+                # FIXME: Ideally there should be a way to specify specific sites
+                # with settings-based credentials. We currently don't, so we can't
+                # use the get_current method that consider sites as well.
+                app = SocialApp.objects.get(provider=provider)
+            else:
+                app = SocialApp(provider=provider)
```

Comment:
This is where you should lean on comments to describe the funkiness of the situation:

```python
# if the SocialApp is defined in settings, it's possible that the database
# model does not yet exist. We are *potentially* going to create it now.
app = SocialApp(provider=provider)
    for field in ["client_id", "secret", "key", "certificate_key"]:
        setattr(app, field, config.get(field))
    # if refresh tokens are not being stored in the database, we won't store
    # the SocialApp in the database either. Doing so would put secrets into the
    # database, and the user who isn't saving refresh tokens might not want
    # that.

    # More importantly, though, it is only necessaray for the SocialApp object
    # to exist in the database when we are going to save refresh tokens,
    # because refresh tokens are foreign-key-related to the SocialApp.
    if app_settings.STORE_TOKENS:
        app.save()
```

As a side note, I'd advocate for avoiding a `FIXME` tag. You're implying a major redesign of the library. It's easy to see in hindsight that there might have been a better design, but a well-described fix is the best thing to do at this time imho.